src/Stratis.Bitcoin.Features.SmartContracts/PoA/Rules/AllowedCodeHashLogic.cs

```diff
@@ -0,0 +1,40 @@
+using NBitcoin;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Features.SmartContracts.Rules;
+using Stratis.SmartContracts.CLR;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.PoA.Rules
+{
+    /// <summary>
+    /// Validates that the hash of the supplied smart contract code is contained in a list of supplied hashes.
+    /// </summary>
+    public class AllowedCodeHashLogic : IContractTransactionValidationLogic
+    {
+        private readonly IWhitelistedHashChecker whitelistedHashChecker;
+        private readonly IContractCodeHashingStrategy hashingStrategy;
+
+        public AllowedCodeHashLogic(IWhitelistedHashChecker whitelistedHashChecker, IContractCodeHashingStrategy hashingStrategy)
+        {
+            this.whitelistedHashChecker = whitelistedHashChecker;
+            this.hashingStrategy = hashingStrategy;
+        }
+
+        public void CheckContractTransaction(ContractTxData txData, Money suppliedBudget)
```

Comment:
this is called from partial validation. it should be called from FV rule
this is due to whitelsited hashes repo being updated in FV so you might end up in a situtaion when you've validated a tx inm let's say block 10. CT is at 2 and at 5 hash is removed. So after block 5 + max reorg (let's say it's 1 for this example) it won't longer be a valid tx.

Reply:
Thanks @noescape00, moved it to a FV rule.